File: apps/desktop/src/services/client/resources/task.ts
import type { Branch, TaskAgentSetting } from "../../../types";
import { call } from "../invoke";
import type { AgentLoopBody, CreateTaskBody, UpdateTaskBody, UpsertTaskBody } from "../types";

export class TaskResource {
  create(ws: string, body: CreateTaskBody) {
    return call<{ id: string }>("task_create", {
      workspaceId: ws,
      title: body.title,
      description: body.description,
    });
  }

  upsert(ws: string, body: UpsertTaskBody) {
    return call<{ id: string; external_id: string; upserted: string }>("task_upsert", {
      workspaceId: ws,
      source: body.source,
      externalId: body.external_id,
      title: body.title,
      description: body.description,
      sourceData: body.source_data,
      url: body.url,
    });
  }

  update(ws: string, id: string, body: UpdateTaskBody) {
    return call<void>("task_update", {
      workspaceId: ws,
      taskId: id,
      title: body.title,
      description: body.description,
      approved: body.approved,
      maxVariants: body.max_variants,
      worktree: body.worktree,
    });
  }

  delete(ws: string, id: string) {
    return call<void>("task_delete", { workspaceId: ws, taskId: id });
  }

  reset(ws: string, id: string) {
    return call<void>("task_reset", { workspaceId: ws, taskId: id });
  }

  branches(ws: string, id: string) {
    return call<Branch[]>("project_task_branches", { workspaceId: ws, taskId: id });
  }

  agentSettings(ws: string, id: string) {
    return call<TaskAgentSetting[]>("task_agent_settings", { workspaceId: ws, taskId: id });
  }

  saveAgentSetting(ws: string, id: string, agentId: string, body: AgentLoopBody) {
    return call<void>("task_agent_setting_set", {
      workspaceId: ws,
      taskId: id,
      agentId,
      amount: body.loop.amount,
    });
  }

  deleteAgentSetting(ws: string, id: string, agentId: string) {
    return call<void>("task_agent_setting_delete", { workspaceId: ws, taskId: id, agentId });
  }
}

File: apps/desktop/src-tauri/src/grpc/task.rs
use modula_client::{
    CreateTask, CreatedTask, ModulaClient, ResetOutcome, UpdateTask, UpsertOutcome, UpsertTask,
};
use modula_types::{Task, TaskAgentSetting};
use serde_json::Value;
use tauri::State;

#[tauri::command]
pub async fn task_list(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
) -> Result<Vec<Task>, String> {
    Ok(engine.list_tasks(&workspace_id).await?)
}

#[tauri::command]
pub async fn task_create(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    title: String,
    description: String,
) -> Result<CreatedTask, String> {
    Ok(engine
        .create_task(CreateTask {
            workspace_id,
            title,
            description: Some(description),
            approved: None,
            max_variants: None,
            worktree: None,
            source_data: None,
        })
        .await?)
}

/// External upsert for the import flow; the engine defaults `synced_at` to
/// today and dedups on `(source, external_id)`.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn task_upsert(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    source: String,
    external_id: String,
    title: String,
    description: Option<String>,
    source_data: Option<Value>,
    url: Option<String>,
) -> Result<UpsertOutcome, String> {
    Ok(engine
        .upsert_task(UpsertTask {
            workspace_id,
            external_id,
            source,
            title,
            description,
            source_data,
            status: None,
            url,
            synced_at: None,
            approved: None,
            max_variants: None,
            worktree: None,
        })
        .await?)
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn task_update(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    title: Option<String>,
    description: Option<String>,
    approved: Option<bool>,
    max_variants: Option<i64>,
    worktree: Option<bool>,
) -> Result<(), String> {
    engine
        .update_task(UpdateTask {
            workspace_id,
            task_id,
            title,
            description,
            approved,
            max_variants,
            worktree,
        })
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn task_delete(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
) -> Result<(), String> {
    engine.delete_task(&workspace_id, &task_id).await?;
    Ok(())
}

#[tauri::command]
pub async fn task_reset(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
) -> Result<ResetOutcome, String> {
    Ok(engine.reset_task(&workspace_id, &task_id).await?)
}

#[tauri::command]
pub async fn task_agent_settings(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
) -> Result<Vec<TaskAgentSetting>, String> {
    Ok(engine.list_agent_settings(&workspace_id, &task_id).await?)
}

#[tauri::command]
pub async fn task_agent_setting_set(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    agent_id: String,
    amount: i64,
) -> Result<(), String> {
    engine
        .set_agent_settings(&workspace_id, &task_id, &agent_id, amount)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn task_agent_setting_delete(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    agent_id: String,
) -> Result<(), String> {
    engine
        .delete_agent_settings(&workspace_id, &task_id, &agent_id)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn variant_update(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    variant_id: String,
    status: Option<String>,
    action: Option<String>,
) -> Result<(), String> {
    engine
        .update_variant(&workspace_id, &task_id, &variant_id, status, action)
        .await?;
    Ok(())
}
